### PR TITLE
Adds support to create/delete metric rollups via dataset-config

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
Signed-off-by: Harkishen-Singh <harkishensingh@hotmail.com>

In relation with https://github.com/timescale/promscale_extension/pull/538 & https://github.com/timescale/o11y-team-applications/issues/228

## Description

This commit adds support for creation or deletion of metric rollups by reading
the dataset-config file. The dataset config file contains a new field `downsample:`
which supports 2 options: `automatic` & `resolution`.

If `automatic` is false, then connector updates this setting in the database (in the `default` table) so that `_prom_catalog.scan_for_new_rollups` does not create any new rollups (Caggs).

`resolution` is a comma separated list of resolutions for downsampling, in a way that
`label:resolution:retention` is the format that is needed.
Example: `short:5m:90d,long:1h:365d`
-> The resolution works in a strict manner. Its aim is to make sure that the database contains
exactly those resolutions that are mentioned in the dataset config. Like, if the database
already contains `short` and `long` downsampled data and the dataset is supplied with `short:5m:90d,very_long:1d:365d`
then `long` downsampled case is deleted and `very_long` is created.
-> `resolution` is applied if `automatic` is `true`. If automatic is not given in dataset config, it defaults to `true`
based on our plan to keep downsampling applied as a default setting.
## Merge requirements

Please take into account the following non-code changes that you may need to make with your PR:

- [ ] CHANGELOG entry for user-facing changes
- [ ] Updated the relevant documentation

### CODE REVIEW

File: pkg/dataset/config.go
Comment (niksajakovljevic) [inline pkg/dataset/config.go:52]: Why don't we add a field here? It will add more structure thus easier to work with `Metrics` struct

File: pkg/dataset/downsample.go
Comment (niksajakovljevic) [inline pkg/dataset/downsample.go:31]: I think having a list of resolutions would be easier to navigate. Strings are tedious to parse/understand and more error prone.
```
type Resolution struct {
   Label string
   Interval time.Duration
   Retention time.Duration
}
```

File: pkg/dataset/downsample.go
Comment (niksajakovljevic) [inline pkg/dataset/downsample.go:58]: Do we validate `automatic: false` and no resolution specified?

File: pkg/rollup/rollup.go
Comment (niksajakovljevic) [inline pkg/rollup/rollup.go:17]: Ok I see you have defined the struct. I'd recommend using it in config struct instead of custom string format.

File: pkg/rollup/rollup.go
Comment (niksajakovljevic) [inline pkg/rollup/rollup.go:26]: This code could go away if we use defined structs in the Downsampling configuration.

**Review by niksajakovljevic — CHANGES_REQUESTED** [review]
I haven't been through all the code yet, but given early feedback so you're not blocked on me. My biggest concern so far is using custom string format for configuration. Maybe there is some reason behind but I'd really like to avoid that if possible. Good work.

File: pkg/runner/client.go
Comment (niksajakovljevic) [inline pkg/runner/client.go:174]: Shouldn't we keep this feature off by default until it's battle tested? Otherwise if there is a bug it will affect all the users who upgrade

File: pkg/rollup/rollup.go
Comment (niksajakovljevic) [inline pkg/rollup/rollup.go:117]: Ok, so here are look for a complement of set B. Can you explain more in comments why this is needed.

File: pkg/rollup/rollup_test.go
Comment (niksajakovljevic) [inline pkg/rollup/rollup_test.go:15]: This would be gone if we are using structs for rollup configuration.

File: pkg/rollup/rollup.go
Comment (niksajakovljevic) [reply pkg/rollup/rollup.go:117]: I think it would be nice mentioning here why we need this.

File: pkg/rollup/rollup.go
Comment (niksajakovljevic) [inline pkg/rollup/rollup.go:61]: It would be good to describe that we want to preserve existing downsampling data if possible to avoid recalculations and only create the new items.

**Comment by niksajakovljevic** [comment]
Went through the rest of the code. Beside my previous comments I've noticed there are no e2e tests to test DB interaction. I'd recommend including those. Otherwise looks good (left few nits about improving some comments).

File: pkg/dataset/downsample.go
Comment (Harkishen-Singh) [reply pkg/dataset/downsample.go:58]: If no resolution is specified (which is the default case), we consider it `5m` & `1h` downsampling case. If a user does not want downsampling, then has to set `automatic: false`.

File: pkg/runner/client.go
Comment (Harkishen-Singh) [reply pkg/runner/client.go:174]: I agree, let's keep it off by default.

File: pkg/rollup/rollup.go
Comment (Harkishen-Singh) [reply pkg/rollup/rollup.go:61]: This is not in line with the design. In the design discussion I recall that we wanted this "strict" behaviour.
We can do what you suggested (ideally I am for it), but for now, maybe we follow what's discussed in the design doc, for things to move quickly.

**Comment by Harkishen-Singh** [comment]
This PR will be merged after the merge of https://github.com/timescale/promscale_extension/pull/538

**Comment by Harkishen-Singh** [comment]
Oops, I just saw that the PR is opened against the master, not the feature branch https://github.com/timescale/promscale/tree/feature_metric_rollup. Need to change this.

File: pkg/util/day_duration.go
Comment (alejandrodnm) [inline pkg/util/day_duration.go:1]: NIT: 

Instead of `util` why don't we move to its own package? You can put it in `internal/dayduration`? Additionally you can change the `DayDuration` to `Duration` because it will be referenced as `dayduration.Duration`. Another option could be `day.Duration`.

I don't like stuffing everything into a `util` or `misc` package. I try to follow https://github.com/golang/go/wiki/CodeReviewComments#package-names

Since we already have one I'd try not to add more things there.

**Pull request merged**
